sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -133,39 +133,90 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
 
   private def visitCompoundBodyImpl(
       ctx: CompoundBodyContext,
-      label: Option[String]): CompoundBody = {
+      label: Option[String],
+      allowVarDeclare: Boolean): CompoundBody = {
     val buff = ListBuffer[CompoundPlanStatement]()
     ctx.compoundStatements.forEach(compoundStatement => {
       buff += visit(compoundStatement).asInstanceOf[CompoundPlanStatement]
     })
 
+    val compoundStatements = buff.toList
+
+    if (allowVarDeclare) {
+      val declareAfterPrefix = compoundStatements
+        .dropWhile(statement => statement.isInstanceOf[SingleStatement] &&
+          statement.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+        .filter(_.isInstanceOf[SingleStatement])
+        .find(_.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+
+      declareAfterPrefix match {
+        case Some(SingleStatement(parsedPlan)) =>
+          throw SqlScriptingErrors.variableDeclarationOnlyAtBeginning(
+            parsedPlan.asInstanceOf[CreateVariable].name.
+              asInstanceOf[UnresolvedIdentifier].nameParts.last,
+            parsedPlan.origin.line.get.toString)
+        case _ =>
+      }
+
+    } else {
+      val declare = compoundStatements
+        .filter(_.isInstanceOf[SingleStatement])
+        .find(_.asInstanceOf[SingleStatement].parsedPlan.isInstanceOf[CreateVariable])
+
+      declare match {
+        case Some(SingleStatement(parsedPlan)) =>
+          throw SqlScriptingErrors.variableDeclarationOnlyAtBeginning(
+            parsedPlan.asInstanceOf[CreateVariable].name.
+              asInstanceOf[UnresolvedIdentifier].nameParts.last,
+            parsedPlan.origin.line.get.toString)
+        case _ =>
+      }
+    }
+
     CompoundBody(buff.toSeq, label)
   }
 
-  override def visitBeginEndCompoundBlock(ctx: BeginEndCompoundBlockContext): CompoundBody = {
-    val beginLabelCtx = Option(ctx.beginLabel())
-    val endLabelCtx = Option(ctx.endLabel())
+  private def generateLabelText(
```

Comment:
nit: indentation for parameters. also, even though it's private function, it's nice to provide return type for function declaration.